xlive/H2MOD/Variants/H2X/H2X.cpp

```diff
@@ -19,11 +27,53 @@ std::vector<H2X::h2x_mod_info> weapons =
 	{ "objects\\weapons\\rifle\\brute_plasma_rifle\\brute_plasma_rifle", 10.0f, 11.0f, 0, true }
 };
 
-// TODO: find a math formula to fix this
+float calculate_h2x_firerate(float h2v_firerate) {
+	// Calculate the effective firerate per seconds.
+	/* This works because every frame Halo 2 checks to see if enough time has
+	   passed for the gun to be allowed to fire again.
+	   These checks happen more often at higher framerates thus where at 30fps
+	   it had to wait an extra frame, it now does not.
+
+	   We simulate it using this formula */
+
+	if (h2v_firerate < 0.0001) {
+		// Do not divide by 0.
+		return h2v_firerate;
+	} else if (h2v_firerate > 30.0) {
+		/* Firing rates above 30 are not possible at 30fps and below. */
+		return 30.0;
+	}
+	/* - 0.1 to avoid a really pesky floating point rounding error
+	   This error and rounding hack only affects higher values above 10.*/
+	return 30.0 / ceil(1 / h2v_firerate * 30.0 - 0.1);
```

Comment:
any reason to declare these values as doubles instead of floats (also that 1 declared as int)? (it could be the root problem with rounding)

Reply:
I have thought about it. Current testing has showed it is only a problem in final builds. Not debug builds, showing that there is an issue with how the math is being optimized by the compiler.

I can make a version of the code with it being converted to a double first.